src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -24605,6 +24605,1499 @@ void Init()
+                // (18,24): warning CS8603: Possible null reference return.
+                //                 return Value; // 1
+                Diagnostic(ErrorCode.WRN_NullReferenceReturn, "Value").WithLocation(18, 24)
+                );
+        }
+
+        [Fact, WorkItem(49750, "https://github.com/dotnet/roslyn/issues/49750")]
+        public void MemberNotNullWhenTrue_SwitchExpressionOnProperty()
+        {
+            var c = CreateCompilation(new[] { @"
+using System.Diagnostics.CodeAnalysis;
+#nullable enable
+
+public struct Test
+{
+    [MemberNotNullWhen(true, nameof(Value))]
+    public bool IsOk { get; }
+    public string? Value { get; }
+
+    public string M()
+    {
+        return IsOk switch
+        {
+            true => Value,
+            _ => throw null!,
+        };
+    }
+}
+", MemberNotNullWhenAttributeDefinition });
+
+            c.VerifyDiagnostics();
+        }
+
+        [Fact, WorkItem(49750, "https://github.com/dotnet/roslyn/issues/49750")]
+        public void MemberNotNullWhenTrue_SwitchExpressionOnMethod()
+        {
+            var c = CreateCompilation(new[] { @"
+using System.Diagnostics.CodeAnalysis;
+#nullable enable
+
+public struct Test
+{
+    [MemberNotNullWhen(true, nameof(Value))]
+    public bool IsOk() { throw null!; }
+    public string? Value { get; }
+
+    public string M()
+    {
+        return IsOk() switch
+        {
+            true => Value,
+            _ => throw null!,
+        };
+    }
+}
+", MemberNotNullWhenAttributeDefinition });
+
+            c.VerifyDiagnostics();
+        }
+
+        [Fact, WorkItem(49750, "https://github.com/dotnet/roslyn/issues/49750")]
+        public void MemberNotNullWhenTrue_SwitchExpressionOnProperty_WhenFalseBranch()
+        {
+            var c = CreateCompilation(new[] { @"
+using System.Diagnostics.CodeAnalysis;
+#nullable enable
+
+public struct Test
+{
+    [MemberNotNullWhen(true, nameof(Value))]
+    public bool IsOk { get; }
+    public string? Value { get; }
+
+    public string M()
+    {
+        return IsOk switch { true => throw null!, _ => Value };
+    }
+}
+", MemberNotNullWhenAttributeDefinition });
+
+            c.VerifyDiagnostics(
+                // (13,16): warning CS8603: Possible null reference return.
+                //         return IsOk switch { true => throw null!, _ => Value };
+                Diagnostic(ErrorCode.WRN_NullReferenceReturn, "IsOk switch { true => throw null!, _ => Value }").WithLocation(13, 16)
+                );
+        }
+
+        [Fact, WorkItem(49750, "https://github.com/dotnet/roslyn/issues/49750")]
+        public void MemberNotNullWhenTrue_Ternary()
+        {
+            var c = CreateCompilation(new[] { @"
+using System.Diagnostics.CodeAnalysis;
+#nullable enable
+
+public class C
+{
+    [MemberNotNullWhen(true, nameof(Value))]
+    public bool IsOk { get; }
+    public string? Value { get; }
+
+    public string M()
+    {
+        return IsOk ? Value : throw null!;
```

Comment:
>throw null! [](start = 30, length = 11)

Consider testing `Value` instead, and merge with test below.

Reply:
The diagnostic produced by a merged test is rather bad to pinpoint what I'm trying to show with these tests:
```
                    // (13,16): warning CS8603: Possible null reference return.
                    //         return IsOk 
                    Diagnostic(ErrorCode.WRN_NullReferenceReturn, @"IsOk 
                ? Value 
                : Value").WithLocation(13, 16)
```

---
In reply to: [573116876](https://github.com/dotnet/roslyn/pull/51001#discussion_r573116876) [](ancestors = 573116876)